lib/raxx/middleware.ex

```diff
@@ -1,6 +1,104 @@
 defmodule Raxx.Middleware do
   alias Raxx.Server
 
+  @moduledoc """
+  A "middleware" is a component that sits between the HTTP server
+  such as as [Ace](https://github.com/CrowdHailer/Ace) and a `Raxx.Server` controller.
+  The middleware can modify requests request before giving it to the controller and
+  modify the controllers response before it's given to the server.
+
+  Oftentimes multiple middlewaress might be attached to a controller and
+  function as a single `t:Raxx.Server.t/0` - see `Raxx.Stack` for details.
+
+  The `Raxx.Middleware` provides a behaviour to be implemented by middlewares.
+
+  ## Example
+
+  Traditionally, middlewares are used for a variety of purposes: managing CORS,
+  CSRF protection, logging, error handling, and many more. This example shows
+  a middleware that given a HEAD request "translates" it to a GET one, hands
+  it over to the controller and strips the response body transforms the
+  response according to the [RFC](https://www.w3.org/Protocols/rfc2616/rfc2616-sec14.html#sec14.13)
+
+  This way the controller doesn't heed to handle the HEAD case at all.
+
+      defmodule Raxx.Middleware.Head do
+        alias Raxx.Server
+        alias Raxx.Middleware
+
+        @behaviour Middleware
+
+        @impl Middleware
+        def process_head(request = %{method: :HEAD}, _config, inner_server) do
```

Comment:
Tempted to suggest the name `next` instead of `inner_server` because it could be server or middleware.

Reply:
Maybe next isn't the best as already used in returns

Reply:
yeah, both server and merchant. I like `inner_server` because it makes it more obvious what it is and what you can do with it.

Reply:
:-D I think middleware (not merchant) was what you were looking for there